### ISSUE TITLE
RotationMatrix

### ISSUE THREAD

**Issue by smichr** [sympy/sympy#24936]
Wondering if Matrices should have the ability to create a RotationMatrix:
```python
def RotationMatrix(axis, angle=None):
    """return 3D rotation matrix for rotations about 'x', 'y', and/or 'z' axis as given in dict or
    2D version if a scalar is given for the axis angle. In 3D, the order of axis defined in the
    dictionary determines the order in which rotations are applied

    EXAMPLES
    ========

    >>> RotationMatrix(pi/4)
    Matrix([[sqrt(2)/2, -sqrt(2)/2], [sqrt(2)/2, sqrt(2)/2]])
    >>> RotationMatrix(dict(x=pi/2,y=pi/4))
    Matrix([
        [ sqrt(2)/2,  sqrt(2)/2, 0],
        [ 0,          0,        -1],
        [-sqrt(2)/2,  sqrt(2)/2, 0]])
    >>> RotationMatrix((-.5,1,1), rad(-74)).n(2)
    Matrix([
        [0.36, 0.48, -0.80],
        [-0.80, 0.60, 0.0015],
        [0.48, 0.64, 0.60]])
    """
    rv = eye(3)
    if angle is None:
      if type(axis) is not dict:
        a = axis
        c = cos(a)
        s = sin(a)
        return Matrix([[c, -s], [s, c]])
      for A in axis:
        a = axis[A]
        c = cos(a)
        s = sin(a)
        i = A.lower()
        if i == 'x':
            rv = Matrix([[1,0,0],[0,c,-s],[0,s,c]])*rv
        elif i == 'y':
            rv = Matrix([[c,0,s],[0,1,0],[-s,0,c]])*rv
        elif i == 'z':
            rv = Matrix([[c,-s,0],[s,c,0],[0,0,1]])*rv
        else:
            assert None, 'unknown axis %s' % A
    else:
        # from https://en.wikipedia.org/wiki/Rotation_matrix
        c = cos(angle)
        o = 1 - c
        s = sin(angle)
        u = x,y,z = Point(*axis).unit.args
        f = lambda x,y,z: x*y*o+z*s
        d = lambda i: c + i**2*o
        return Matrix([
            [d(x),      f(x,y,-z), f(x,z,y)],
            [f(x,y,z),  d(y),      f(y,z,-x)],
            [f(x,z,-y), f(z,y,x),  d(z)]])
    return rv
```

**Comment by samithkavishke** [comment]
may I try this issue

**Comment by smichr** [comment]
It might be good to wait to see if there is some support for this. SymPy doesn't need *all* batteries. But my initial hunch is that this is a pretty common one to want to have if you are working with geomoetry or mechanics. (I haven't looked to see if this is already hiding somewhere in the physics directory: that might be a good place to shart.)

**Comment by sylee957** [comment]
We have refined some rotation matrix api
https://docs.sympy.org/dev/modules/matrices/matrices.html#rotation-matrices

**Comment by evbernardes** [comment]
I'd argue a `Matrix.from_axis_angle` in the same fashion as `Quaternion.from_axis_angle` in which the `axis` parameter would be a vector could be more useful.

For a rotation from common Euler angles, there's always the functions listed by @sylee957.

**Comment by evbernardes** [comment]
What I also find would be nice if there's any interest: a class for Euler angles that would have all of the functions relating to angles. Something like:

```
seq = 'zyx'
angles = [pi/2, 0, pi/2]
euler = Euler(seq, angles)
rot = euler.to_rotation_matrix()
```
I intend on working on this as a side project separated from sympy at some point, though, in order not to polute sympy with extra code...

**Comment by moorepants** [comment]
It may be worth noting that we have all Euler and Tait-Bryan rotation matrices and angular velocity vectors implemented in the mechanics package (as well as arbitrary single axis rotations and quaternions). I'm reasonably confident that they are correct implementations since we've been using them for over a decade. Feel free to use for inspiration, copying, or correctness checks.

**Comment by smichr** [comment]
> For a rotation from common Euler angles, there's always the functions listed

Yes, but one has to remember to apply them in the right to left order and negate the angles to follow a right hand rule ccw rotation.

I like the idea of a  Euler object. But I would suggestion just calling it `Rotation(angle, axis='z')` and allow it to change to Quaternion or Matrix as desired. A rotation could be defined by a scalar or a vector of length up to 3 with equal length `axis` telling which order the angles are applied, e.g. `Rotation((1,2,3,4),'xyxz')`

>  Euler and Tait-Bryan rotation matrices

@moorepants , maybe I am missing something in my search but I only saw rotation matrices with a lot of extra mechanics decoration (use of frames, etc) to generate a matrix.

**Comment by evbernardes** [comment]
> I like the idea of a Euler object. But I would suggestion just calling it `Rotation(angle, axis='z')` and allow it to change to Quaternion or Matrix as desired. A rotation could be defined by a scalar or a vector of length up to 3 with equal length `axis` telling which order the angles are applied, e.g. `Rotation((1,2,3,4),'xyxz')`

I like a lot how the folks from `Scipy` implemented their `Rotation` class: it's saved as a quaternion, but that's invisible to the user. The way to interact with it is from their `from_XXX` and `as_XXX` methods, in which `XXX` can be `quat`, `euler`, etc... But, of course, this might make more sense in a numerical module.

I'd say that the problem with having a class called `Rotation` that expects Euler angles is that it would be a bit ambiguous, since Rotations can be parametrized in so many ways (and most of them, maybe all, are usually more efficient than Euler angles)

**Comment by evbernardes** [comment]
> Yes, but one has to remember to apply them in the right to left order...

I forgot to comment on this! Not only right-to-left but also, possibly, left-to-right in case of intrinsic angles (which are more commonly used). Which is very confusing...

> ...and negate the angles to follow a right hand rule ccw rotation.

Apart from the original `Matrix.rot_axis1`, `Matrix.rot_axis2` and `Matrix.rot_axis3`, I implemented `Matrix.rot_ccw_axis1`, `Matrix.rot_ccw_axis2` and `Matrix.rot_ccw_axis3`. I had opened an issue (see #24529) to point out that it's confusing that the standard used in the implementations is the clockwise instead of counter-clockwise.

I think that it would be best to deprecate the originals (`Matrix.rot_axisX`) in favor of more explicit names (`Matrix.rot_cw_axisX` and `Matrix.rot_ccw_axisX`).

**Comment by evbernardes** [comment]
If you choose the most commonly used standard of the right hand rule, though, it is at least supposed to be well defined ([see this](https://en.wikipedia.org/wiki/Right-hand_rule)):


![Right-hand_grip_rule](https://user-images.githubusercontent.com/15084103/229162607-ca2f9c4c-06b5-48c6-aa45-f692612650f9.svg)

A good thing about having `cw` and `ccw` as the only options, though, is that it would make the user to choose between them, forcing them to think about what they really want.

**Comment by smichr** [comment]
I think we can avoid more confusion by using "looking direction and ccw/cw" or "hand rule and positive rotation" and not using a mixed metaphor, so to speak, by mixing cw with rule (per comment [here](https://github.com/sympy/sympy/pull/25068/files#r1172062021)).

**Comment by moorepants** [comment]
> @moorepants , maybe I am missing something in my search but I only saw rotation matrices with a lot of extra mechanics decoration (use of frames, etc) to generate a matrix.

All of our methods (e.g. https://docs.sympy.org/latest/modules/physics/vector/api/classes.html?highlight=referenceframe#sympy.physics.vector.frame.ReferenceFrame.orient_body_fixed) to generate rotation matrices are associated with our ReferenceFrame class.
